store/stats_pusher.go

```diff
@@ -2,229 +2,84 @@ package store
-}
-
-type noopStatsPusher struct{}
-
-func (noopStatsPusher) Start() error { return nil }
-func (noopStatsPusher) Close() error { return nil }
-func (noopStatsPusher) Send([]byte)  {}
-
-type websocketStatsPusher struct {
-	boot    *sync.Mutex
-	conn    *websocket.Conn
-	cancel  context.CancelFunc
-	send    chan []byte
-	sleeper utils.Sleeper
-	started bool
-	url     url.URL
-}
-
-// NewWebsocketStatsPusher returns a stats pusher using a websocket for
-// delivery.
-func NewWebsocketStatsPusher(url *url.URL) StatsPusher {
-	return &websocketStatsPusher{
-		url:     *url,
-		send:    make(chan []byte, 100), // TODO: figure out a better buffer (circular FIFO?)
-		boot:    &sync.Mutex{},
-		sleeper: utils.NewBackoffSleeper(),
+	return &StatsPusher{
+		ORM:      orm,
+		WSClient: wsClient,
+		Period:   60 * time.Second,
 	}
 }
 
-// Start starts a write pump over a websocket.
-func (w *websocketStatsPusher) Start() error {
-	w.boot.Lock()
-	defer w.boot.Unlock()
-
-	if w.started {
-		return nil
+// Start starts the stats pusher
+func (eq *StatsPusher) Start() error {
+	err := eq.WSClient.Start()
+	if err != nil {
+		return err
 	}
-
 	ctx, cancel := context.WithCancel(context.Background())
-	w.cancel = cancel
-	wg := &sync.WaitGroup{}
-	wg.Add(1)
-	go w.connectAndWritePump(ctx, wg)
-	wg.Wait()
-	w.started = true
+	eq.cancel = cancel
+	go eq.pollEvents(ctx)
 	return nil
 }
 
-const (
-	// Time allowed to write a message to the peer.
-	writeWait = 10 * time.Second
-
-	// Time allowed to read the next pong message from the peer.
-	pongWait = 60 * time.Second
-
-	// Send pings to peer with this period. Must be less than pongWait.
-	pingPeriod = (pongWait * 9) / 10
-
-	// Maximum message size allowed from peer.
-	maxMessageSize = 512
-)
-
-// Inspired by https://github.com/gorilla/websocket/blob/master/examples/chat/client.go
-// lexical confinement of done chan allows multiple connectAndWritePump routines
-// to clean up independent of itself by reducing shared state. i.e. a passed done, not w.done.
-func (w *websocketStatsPusher) connectAndWritePump(parentCtx context.Context, wg *sync.WaitGroup) {
-	wg.Done()
-	logger.Info("Connecting to linkstats at ", w.url.String())
-
-	for {
-		select {
-		case <-parentCtx.Done():
-			return
-		case <-time.After(w.sleeper.After()):
-			connectionCtx, cancel := context.WithCancel(parentCtx)
-			defer cancel()
-
-			if err := w.connect(connectionCtx); err != nil {
-				logger.Warn("Failed to connect to linkstats (", w.url.String(), "): ", err)
-				break
-			}
-
-			logger.Info("Connected to linkstats at ", w.url.String())
-			w.sleeper.Reset()
-			go w.readPumpForControlMessages(cancel)
-			w.writePump(connectionCtx)
-		}
+// Close shuts down the stats pusher
+func (eq *StatsPusher) Close() error {
+	if eq.cancel != nil {
+		eq.cancel()
 	}
+	return eq.WSClient.Close()
 }
 
-var (
-	newline = []byte{'\n'}
-)
+func (eq *StatsPusher) pollEvents(parentCtx context.Context) {
+	pollTicker := time.NewTicker(eq.Period)
```

Comment:
Not for this PR as discussed but I'm excited to see if we can elegantly get off polling. There are implications with integration testing if we have to wait one minute for a sync event.